### ISSUE TITLE
hardcoded signal number in systemd-exit.service.in is not portable

### ISSUE THREAD

**Issue by dtzWill** [systemd/systemd#8621]
### Submission type

 - Bug report

### systemd version the issue has been seen with

v238, master

### Summary

Value of `SIGRTMIN+24` can only be determined at runtime, but systemd assumes the value is `58` statically,.

### Code
Relevant code is here: https://github.com/systemd/systemd/blob/a7df2d1e4349a8c90223d8b7ae339dbad05da503/units/user/systemd-exit.service.in#L19

### Impact and Background

Breaks systemd on platforms / environments that have different values for `SIGRTMIN+24`.  Man page for signal suggests this might not be portable across glibc versions either[1].  Same man page suggests this value should never be hard-coded:

       The Linux kernel supports a range of 33 different real-time signals,
       numbered 32 to 64.  However, the glibc POSIX threads implementation
       internally uses two (for NPTL) or three (for LinuxThreads) real-time
       signals (see pthreads(7)), and adjusts the value of SIGRTMIN suitably
       (to 34 or 35).  Because the range of available real-time signals
       varies according to the glibc threading implementation (and this
       variation can occur at run time according to the available kernel and
       glibc), and indeed the range of real-time signals varies across UNIX
       systems, programs should never refer to real-time signals using hard-
       coded numbers, but instead should always refer to real-time signals
       using the notation SIGRTMIN+n, and include suitable (run-time) checks
       that SIGRTMIN+n does not exceed SIGRTMAX.

Getting this wrong causes failure to cleanly exit user sessions.

### Proposed Fixes

1) Require `kill` supports notation for expressing the signal relative to `SIGRTMIN`.  The commit introducing this context mentions utillinux, but this appears to no longer be the case.  A quick check shows support was added in 2011: https://github.com/karelzak/util-linux/commit/ae63d823abc33554ee1d3906a876734277293557

2) Generate the constant using compile-time check (print result with a small program or so).  While technically this value can only be determined at runtime, for  practical purposes this value is really a per-libc constant and would likely be sufficient to address portability concerns.  If it's not safe to assume it is constant then many things would break in worse ways :).

[1] http://man7.org/linux/man-pages/man7/signal.7.html

**Comment by yuwata** [comment]
May be related to #7001.

**Comment by poettering** [comment]
we don't support LinuxThreads-based glibcs in systemd really... Is this a theoretic issue, or do you actually have a real-life system where this matters?

The way I read that is that people who care about portability to other UNIXes should use SIGRTMIN/SIGRTMIN but as long as you stick to current glibc/Linux things should be stable... and we aren't portable to non-Linux anyway....

That said, either way, if currentl util-linux and procps supports the SIGRTMIN syntax just fine, we should really fix this.

BTW, we used to use SIGRTMIn actually, and switched this over:

https://github.com/systemd/systemd/commit/5411ae4782270e8e82b05e4c7c8135f4eb774dda

Back then some distros used procps's kill, others used util-linux' kill, and iirc the procps one didn't grok SIGRTMIN. does it now?

**Comment by filbranden** [comment]
> Back then some distros used procps's kill, others used util-linux' kill, and iirc the procps one didn't grok SIGRTMIN. does it now?

procps-ng's kill does and I believe that's what's shipping in modern versions of distros...

**Comment by floppym** [comment]
> as long as you stick to current glibc/Linux things should be stable

There have been some weird archs (like HPPA) that steal additional RT signals, so even glibc/NPTL isn't entirely stable. See 4dffec1459f50ac9f8f67ccfcb79836b4ed5a50e and 8750ac02381be5f0b3ebbcc0131779553fa7ce52.

**Comment by floppym** [comment]
I verified that `kill` from procps-3.3.12 and GNU coreutils-8.29 both support the -s `SIGRTMIN+n` syntax, so it should be pretty safe to use it now.

**Comment by evverx** [comment]
In general, I think it is a good idea to send `RT*` signals. In practice, it is possible to build `busybox` without `FEATURE_RTMINMAX`; in this case `kill` will reject any name beginning with `[SIG]RT*`:
```sh
$ busybox kill -s SIGRTMIN+1 1
kill: bad signal name 'SIGRTMIN+1'
```
so it would probably make sense to fall back to the numerical signal if `kill -s RT*` fails. Something like `/bin/sh -c '@KILL@ -SIGRTMIN+24 $MANAGERPID || @KILL@ -s 58 $MANAGERPID` should probably work in most cases.

**Comment by dtzWill** [comment]
> In general, I think it is a good idea to send RT* signals. In practice, it is possible to build busybox without FEATURE_RTMINMAX; in this case kill will reject any name beginning with [SIG]RT*:

Maybe, but it seems simpler to "require" a `kill` binary that can handle this.  Especially if it's the case that most every implementation does unless you go out of your way to disable it.  Even so, we can check if it does at compile-time, at least in the non-cross case.

----------

> The way I read that is that people who care about portability to other UNIXes should use SIGRTMIN/SIGRTMIN but as long as you stick to current glibc/Linux things should be stable... and we aren't portable to non-Linux anyway....

From that manpage, and from glibc's source (see references below) I'm not sure that's in any way something we can rely upon-- it seems rather clear they don't consider `SIGRTMIN` a constant and repeatedly advise against treating it as such.

However I doubt they'd make such a change any time soon, if ever, so take that as you will :+1:.

----------

This is not a purely theoretical issue, FWIW, I encountered it when using systemd w/musl.
That work is ongoing, largely based on OpenEmbedded's work, but works well enough now (after working around this issue) that other parts of my project are what needs attention not anything regarding systemd :).

----------

See also:

### glibc's source

* [bits/signum-generic.h](https://sourceware.org/git/?p=glibc.git;a=blob;f=bits/signum-generic.h;h=8f1044a14479b97b491e0de6228a767067ba2614;hb=688903eb3ef01301d239ab753d309d45720610a7#l90)
* [signal/allocrtsig.c](https://sourceware.org/git/?p=glibc.git;a=blob;f=signal/allocrtsig.c;h=fa642a52eac7c6c3cbb89dc681215cd7e38f5168;hb=688903eb3ef01301d239ab753d309d45720610a7#l27)
* NPTL (FWIW): [sysdeps/nptl/allocrtsig.c](https://sourceware.org/git/?p=glibc.git;a=blob;f=sysdeps/nptl/allocrtsig.c;h=2039d62a83d9c5e895c833a03a28d5642d92d364;hb=688903eb3ef01301d239ab753d309d45720610a7#l22)

Particularly of interest is the difference between `__SIGRTMIN` and `SIGRTMIN`--which I don't suggest using at all (although apparently qemu does!) but demonstrates how they believe folks should be using `SIGRTMIN`.

**Comment by evverx** [comment]
It seems simpler not to change anything until we somehow can be sure that the change doesn't create more problems than it solves. Basically, I was just trying to show that it is possible in 2018 to have `kill` that doesn't support `[SIG]RT*`. In my opinion, this is far less theoretical than the use case you described above.

**Comment by dtzWill** [comment]
> It seems simpler not to change anything until we somehow can be sure that the change doesn't create more problems than it solves. Basically, I was just trying to show that it is possible in 2018 to have kill that doesn't support [SIG]RT*. In my opinion, this is far less theoretical than the use case you described above.

Well since I've actually encountered this issue, I of course disagree on what is "less" theoretical.

Thinking about your point some more, I think to the extent it's "reasonable" and "practical" to provide systemd a version of the `kill` utility that does not understand real-time signal names.... this does not motivate or defend hard-coding the value of `SIGRTMIN+24` but rather suggests that relying on user-provided utility to have certain behavior is fragile.

This is the current situation:

"When building systemd, please provide a binary (and ensure it is available at runtime at the indicated path) that when invoked with arguments `-s 58 pid` sends signal `SIGRTMIN+24` to the process with id `pid`."

Now this is not written anywhere, but is the requirement for correct behavior.  Such a utility would be easy to write in a correct way, but it does seem like a strange requirement.

Instead systemd asks for the path of the `kill` command, and hopes that when it says `-s 58` that the appropriate signal will be issued-- even though such an assumption is not at all guaranteed by the request for `kill`  (and is repeatedly explicitly discouraged by the platform documentation).

Indeed, it does seem silly to rely on such a utility at all-- when what we really want is `kill(pid, SIGRTMIN+24)`-- and already have code that does so properly from a variety of contexts.

A quick scan of the source (please confirm) suggests that the only use of the `kill` path is to perform this exact behavior... if true I think it makes it very clear that a runtime dependency on an external utility is unnecessary.

I am not an expert on the workings and code of systemd but it seems if what we want is to issue the signal `SIGRTMIN+24` then there are a great many ways that this could be accomplished.

Two directions for this:

1) Provide our own entry point that has the desired functionality, so that it can be invoked in the `ExecStart` of the `systemd-exit` user service.  A quick look around suggests we're already close with `systemctl kill -s 'RTMIN+24' target`, although rolling a new utility for this would be trivial (leveraging existing systemd signal code).

2) Architect around the use of a "service" that "executes" by sending a signal.  I'm not saying this is bad but it does seem like something that could be achieved in other ways that might be cleaner or using 
 new abstraction useful beyond this particular need.

Either way would help make things a bit cleaner in one way or another, and while it may seen to be "okay" on all platforms systemd is interested in I'll note:

* Just because it works today doesn't mean it will work tomorrow, if it's not in a standard or other agreement.
* ... especially if the relevant standards and documentation go out of their way to advise against assuming it won't work differently.

(here's your "theoretical": I think it would be compliant--although either useless or extreme-- to change value of `SIGRTMIN` across processes and executions of the same process... perhaps some sort of "signal randomization" hardening scheme? who knows :))

---------------

Point about avoiding creating more problems than are solved is a good one, and I agree wholeheartedly :).

**Comment by evverx** [comment]
It seems that the word "theoretical" wasn't correct in that context. I should have used the word "probability" or something similar. I'm sorry about that.

> Provide our own entry point that has the desired functionality

I think it would be great if `systemctl -f -f kill --signal=SIGRTMIN+24 PID` could be used to send signals to processes directly bypassing `systemd` in a way similar to `systemctl -ff poweroff` calling `reboot` without contacting `systemd`. It's not the best UI I've ever seen, though :-)

**Comment by dtzWill** [comment]
> It seems that the word "theoretical" wasn't correct in that context. I should have used the word "probability" or something similar. I'm sorry about that.

Thank you, sorry if I misinterpreted!

> I think it would be great if systemctl -f -f kill --signal=SIGRTMIN+24 PID could be used to send signals to processes directly bypassing systemd in a way similar to systemctl -ff poweroff calling reboot without contacting systemd. It's not the best UI I've ever seen, though :-)

Yeah! That sort of thing would be great and probably useful for other reasons :).

------

## Whatever we do, don't use `kill -s 'RTMIN+24'`

Sensationalist header aside, after some (more) investigation I think this would be a big mistake and indeed cause many problems on platforms/environments of many users.  Here's what I found:

`RTMIN` as a constant is interpreted differently among implementations of `kill`.
On reflection, this is not surprising -- it arises from the classic problem of every component having a different idea of what belongs on which side of the "system" interface (anyone have a name for this problem?)-- but is made worse here in that the string must somehow encode "which" "RTMIN" is intended--that of the kernel? Or that of the libc? (or that of the debugging runtime that reserves a few more signals, or that of the emulator wrapping the code, etc.)

More concretely, I grabbed copies of busybox, coreutils, utillinux, and procps-ng:

```console
$ for x in busybox coreutils utillinux procps-ng; do echo "-------"; echo $x; nix run -f channel:nixos-unstable strace $x -c strace -e execve,kill kill -s 'RTMIN+24' 1; done
-------
busybox
execve("/nix/store/7dk9vv9ns2sp8r1xhwwf3ahj9k2yg1a9-busybox-1.28.1/bin/kill", ["kill", "-s", "RTMIN+24", "1"], 0x7ffeac1c6b38 /* 66 vars */) = 0
kill(1, SIGRT_24)                       = -1 EPERM (Operation not permitted)
kill: can't kill pid 1: Operation not permitted
+++ exited with 1 +++
-------
coreutils
execve("/nix/store/qrxs7sabhqcr3j9ai0j0cp58zfnny0jz-coreutils-8.29/bin/kill", ["kill", "-s", "RTMIN+24", "1"], 0x7fff174bb788 /* 66 vars */) = 0
kill(1, SIGRT_26)                       = -1 EPERM (Operation not permitted)
kill: ‘1’: Operation not permitted
+++ exited with 1 +++
-------
utillinux
execve("/nix/store/54gdanvd9mzzwymvsxw02vwhj4h596hw-util-linux-2.31.1-bin/bin/kill", ["kill", "-s", "RTMIN+24", "1"], 0x7ffd31e17a38 /* 66 vars */) = 0
kill(1, SIGRT_26)                       = -1 EPERM (Operation not permitted)
kill: sending signal to 1 failed: Operation not permitted
+++ exited with 1 +++
-------
procps-ng
execve("/nix/store/r3f0c829y6rvv15s8b1slkci6i13b6yx-procps-3.3.12/bin/kill", ["kill", "-s", "RTMIN+24", "1"], 0x7ffdcef02ab8 /* 66 vars */) = 0
kill(1, SIGRT_26)                       = -1 EPERM (Operation not permitted)
kill: (1): Operation not permitted
+++ exited with 1 +++
```

(paths shown in case anyone cares about versions or build logs/parameters or whatever)

In short, it seems busybox interprets `RTMIN` different than the rest, which the source code explains is due to their use of `__SIGRTMIN` when available (glibc-only, as far as I know).

I don't believe there is a notation that clarifies this, although I didn't look too closely.

--------
tl;dr: busybox interprets "RTMIN" differently than other implementations, so we probably shouldn't rely on the ability for a user-supplied "kill" utility to have any particular functionality (without testing or something).

**Comment by evverx** [comment]
Thank you for sharing this. I was going to file an issue on April 1 about [`systemd` not receiving `SIGKILL`](https://github.com/torvalds/linux/commit/628c1bcba204052d19b686b5bac149a644cdb72e) and [`SIGSTOP`](https://github.com/torvalds/linux/commit/2d39b3cd34e6d323720d4c61bd714f5ae202c022). Now I don't think anything related to sending signals is funny :-)

**Comment by evverx** [comment]
After I had been thinking for a while about how to add to the documentation a meaningful explanation of what `systemctl -ff kill PID` would do and why it wouldn't allow people to specify the names of services, I changed my mind. I don't think `systemctl kill` should be used to send raw signals. Perhaps it would be better to add another subcommand named `kill-pid` which would probably make it more intuitive. Having said that, I believe that all the implementations of `kill` get `RTMAX` right, so I'm tempted to say that I personally would be happy with:
```
/bin/sh -c '@KILL@ -s RTMAX-6 $MANAGERPID || @KILL@ -s 58 $MANAGERPID
```
though I understand that it's ugly and `/bin/sh` has always been avoided or removed.

**Comment by evverx** [comment]
I also think that [the "Signals" section of the documentation](https://www.freedesktop.org/software/systemd/man/systemd.html#Signals) might be a little bit confusing because, as it turns out, nobody agrees on what `RTMIN` means and perhaps it is worth mentioning what `RTMIN` means there to simplify things a bit.

**Comment by dtzWill** [comment]
As a PoC of sorts, this introduces a "kill-user-manager" alias for this purpose: https://github.com/dtzWill/systemd/commit/1dcda48399b549193f62a265e4f9199b768200e0

Using a different entry point or plumbing through a sub-command might be a better approach, but thought I'd share anyway.

**Comment by evverx** [comment]
I'd prefer it if the alias weren't introduced because people might start using it for whatever reason and it will be impossible to get rid of it without breaking someone's weird setup.

**Comment by evverx** [comment]
The previous comment is based on the assumption that one day we will be able to send `RT*` signals via `kill`, which will make the stopgap unnecessary. If the assumption doesn't look realistic or it isn't considered a stopgap, it probably doesn't matter much which method we choose to get around the inconsistency in `kill`, but I think the fewer moving parts we have, the better.

**Comment by medhefgo** [comment]
Let me suggest something here: Why do we need to fork+exec to send a signal to a process in the first place? It seems wasteful when systemd could just call kill() by itself. Afaik, it doesn't block so it should be safe to do, no? Maybe a SendSignal= directive of sorts is in order?

**Comment by evverx** [comment]
I'm adding the "needs-patch" label because the issue seems to mainly has to do with `musl`, where `RTMIN` is always `35` and `SIGRTMAX` has the wrong value on `mips`, which apparently prevents the use of `kill -s RTMAX-6`.

**Comment by evverx** [comment]
I think that an additional directive to get around a way in which `musl`, which is unsupported, by the way, deal with `RTMIN` and  `RTMAX` is a bit too much.

**Comment by evverx** [comment]
`RTMAX-6` doesn't seem to be portable either. I don't know what I was thinking about when I was suggesting it. Sorry for the noise.

**Comment by poettering** [comment]
> In general, I think it is a good idea to send RT* signals. In practice, it is possible to build busybox without FEATURE_RTMINMAX; in this case kill will reject any name beginning with [SIG]RT*:

we really don't design our stuff for arbitrary combinations of stuff. And I am very sure that busybox with rt sig support turned off is far outside of what we need to care for.

**Comment by poettering** [comment]
> There have been some weird archs (like HPPA) that steal additional RT signals, so even glibc/NPTL isn't entirely stable. See 4dffec1 and 8750ac0.

well, the hppa thing just means different numbers of rt sigs, but it doesn't change their offset. that means SIGRTMIN+ should be stable, even though SIGRTMAX- might not be. However, the case here is a SIGRTMIN+ one

**Comment by poettering** [comment]
So, we document "util-linux" as a dep of systemd anyway. If people ignore that and pull in busybox, it's fine if they do, but they are on their own anyway, we don't need to care.

If busybox defines SIGRTMIN differently from everybody else, then that's a bug in busybox simply, they need to fix it. They don't get to redefine what SIGRTMIN means, that's between the libc and the kernel. We don't work around buggy software in our code. It's the wrong place. 

Now, that all said, I think we should actually fix this a different way: the --system version of "systemd-exit.service" already uses "systemctl --force exit", instead of "kill". We should make the same work for the "--user" instance, i.e. make "systemctl --user --force exit" work and use that. That's not only prettier and more systematic, but makes the whole problem go away.

@medhefgo and yes, you have a point, it's a bit weird to fork out a process from our manager whose only job is to tell the manager to do something. Maybe it would indeed be better to shortcut that. I figure we could introduce StartAction= or so, which builds on FailureAction=/SuccessAction= and so on, but is actually the main operation of a "start" invocation of the unit.

**Comment by evverx** [comment]
@poettering `systemd --user --force` sounds great.

>  So, we document "util-linux" as a dep of systemd anyway.

I'm a bit confused about the commit where `kill` from `propcps` was mentioned. What was the idea behind it if the only supported `kill` is from `util-linux`?

**Comment by poettering** [comment]
> I'm a bit confused about the commit where kill from propcps was mentioned. What was the idea behind it if the only supported kill is from util-linux?

Well, true, but what I wanted to say is that you can't build systemd without util-linux anyway, hence you'll have its version of kill around anyway, hence we kinda depend on a "working" kill version anyway, hence replacing it with a worse one, that doesn't understand SIGRTMIN properly then that's extra work for them they have to deal with. Or to say this differently: we explicitly declare deps in our README on an implementation that is "at least as good" as we need it. If people fall back below that, it's on them. procps' kill implementation otoh is fine these days, it's not a regression for the explicit dep we declare.

**Comment by evverx** [comment]
I see. Thank you. I wouldn't say it's worse, though. It's just different :-)

**Comment by poettering** [comment]
> I see. Thank you. I wouldn't say it's worse, though. It's just different :-)

well, given that it apparently uses a broken SIGRTMIN definition I'd say it's actually worse, because simply broken ;-)

**Comment by floppym** [comment]
> well, the hppa thing just means different numbers of rt sigs, but it doesn't change their offset. that means SIGRTMIN+ should be stable, even though SIGRTMAX- might not be. However, the case here is a SIGRTMIN+ one

systemd-exit.service currently hard-codes an absolute signal number (58), NOT a value relative to `SIGRTMIN`. That's what this issue is all about. ;)

**Comment by poettering** [comment]
> well, the hppa thing just means different numbers of rt sigs, but it doesn't change their offset. that means SIGRTMIN+ should be stable, even though SIGRTMAX- might not be. However, the case here is a SIGRTMIN+ one
> 
> systemd-exit.service currently hard-codes an absolute signal number (58), NOT a value relative to SIGRTMIN. That's what this issue is all about. ;)

yes, precisely, but the signal it means by that is documented and implemented by its `SIGRTMIN+xxx` name (and not by its `SIGRTMIN-xxx`) name, end hence hence differences on hppa do not affect what `58` is supposed to mean
